Pegasus/include/Geometry.hpp

```diff
@@ -328,107 +354,472 @@ struct Cache<Box, Box> : CacheBase
+
+    return cache->hp.RayIntersection(cache->rayNormal, cache->rayOrigin, cache->intersection);
+}
+
+/** Ray, Plane CalculateContactNormal specialization */
+template <>
+inline glm::dvec3 CalculateContactNormal<Ray, Plane>(SimpleShape const* a, SimpleShape const* b, IntersectionCacheBase* cacheBase)
+{
+    auto cache = static_cast<IntersectionCache<Ray, Plane>*>(cacheBase);
+
+    return cache->rayNormal;
+}
+
+/** Ray, Plane CalculatePenetration specialization */
+template <>
+inline double CalculatePenetration<Ray, Plane>(SimpleShape const* a, SimpleShape const* b, IntersectionCacheBase* cacheBase)
+{
+    return std::numeric_limits<double>::max();
+}
+
+/** Ray, Sphere Initialize specialization */
+template <>
+inline void Initialize<Ray, Sphere>(SimpleShape const* a, SimpleShape const* b, IntersectionCacheBase* cacheBase)
+{
+    auto cache = static_cast<IntersectionCache<Ray, Sphere>*>(cacheBase);
+    auto ray = static_cast<Ray const*>(a);
+    auto sphere = static_cast<Sphere const*>(b);
+
+    cache->rayOrigin = ray->centerOfMass;
+    cache->rayNormal = ray->direction;
+    cache->sphereRadius = sphere->radius;
+    cache->sphereCenter = sphere->centerOfMass;
+
+    glm::dvec3 const L = cache->sphereCenter - cache->rayOrigin;
+    double const tc = glm::dot(L, cache->rayNormal);
+    double const dd = glm::dot(L, L) - tc * tc;
+    cache->intersection = dd > 0;
+
+    if (cache->intersection)
+    {
+        double const d = glm::sqrt(dd);
+        double const t1c = glm::sqrt(glm::pow2(cache->sphereRadius) - d * d);
+        double const t1 = tc - t1c;
+        double const t2 = tc + t1c;
+        cache->intersection = (cache->sphereRadius - d) >= 0.0;
+        cache->inPoint = cache->rayOrigin + cache->rayNormal * t1;
+        cache->outPoint = cache->rayOrigin + cache->rayNormal * t2;
+        cache->penetration = glm::length(cache->inPoint - cache->outPoint);
+        cache->sphereContactNormal = glm::normalize(cache->inPoint - cache->sphereCenter);
+    }
+}
+
+/** Ray, Sphere CalculateIntersection specialization */
+template <>
+inline bool CalculateIntersection<Ray, Sphere>(SimpleShape const* a, SimpleShape const* b, IntersectionCacheBase* cacheBase)
+{
+    auto cache = static_cast<IntersectionCache<Ray, Sphere>*>(cacheBase);
+
+    return cache->intersection;
+}
+
+/** Ray, Sphere CalculateContactNormal specialization */
+template <>
+inline glm::dvec3 CalculateContactNormal<Ray, Sphere>(SimpleShape const* a, SimpleShape const* b, IntersectionCacheBase* cacheBase)
+{
+    auto cache = static_cast<IntersectionCache<Ray, Sphere>*>(cacheBase);
+
+    return cache->sphereContactNormal;
+}
+
+/** Ray, Sphere CalculatePenetration specialization */
+template <>
+inline double CalculatePenetration<Ray, Sphere>(SimpleShape const* a, SimpleShape const* b, IntersectionCacheBase* cacheBase)
+{
+    auto cache = static_cast<IntersectionCache<Ray, Sphere>*>(cacheBase);
+
+    return cache->penetration;
```

Comment:
Why is this value calculated in the `Initialize` function? Wouldn't it make more sense to calculate it in this function? Same for `CalculateContactNormal` and `CalculateIntersection` in this case.